### ISSUE TITLE
Concept: Configurable Event Logging

### ISSUE THREAD

**Issue by ejoerns** [rauc/rauc#1070]
So far, RAUC provides a status file where information about the latest installation is stored.
This is useful to know what is currently on the system but does not (or only partly) answer common questions such as:

* What has been installed on this device before?
* How many times did the installed system boot (successfully)?
* When has which redundancy group been activated?
* Did we run into a fallback?

A proper solution to gain more history-related information would be to establish a dedicated logging mechanism in RAUC.

Since the actual events to log might differ between different use cases as well as the logging policy may differ, we propose a configurable event logging framework in RAUC here.

The key idea is to have different loggers configurable in the system configuration file. Each logger can specify the events to log and where to write the log. A good default location for logging would be a file in RAUC's data-dir.
Additional options may define log rotation handling, compression, or fsync() handling.
Default output should be a human-readable format, but machine-readable formats may follow.

An initial draft of system.conf sections might look as follows:

``` ini
[log.boot]
filename=boot.log # relative in data-dir
events=booted-version
max-size=1M
max-files=5
compress="xz -9"

[log.install]
filename=/srv/rauc/install.log
events=install
sync=true # after each write: flush und fsync
max-size=0 # no size limit
max-files=0 # no rotation
```

Internally an event log message will contain meta information that can be used for printing, filtering, etc.
As these event log messages will replace parts of the current log messages, we can use the additional meta information to also extend the information sent to the systemd-journal, etc. E.g. with using a message [catalog](https://www.freedesktop.org/wiki/Software/systemd/catalog/) for providing more details.

The high-level structure of a log item might look as follows:

  * Log item:
      * timestamp
      * event type
      * event ID  (MESSAGE_ID=$(systemd-id128 new))
      * payload

A failure in logging should not be fatal for the installation process but might disable the corresponding log handler.

Two pieces of information that will be relevant for answering parts of the above-noted questions from looking at the logs are still missing in RAUC and need to be added:

**Transaction ID:** A unique ID that allows unambiguously identifying a specific installation process. It will change for each install invocation and might either be provided by the user or will be auto-generated if not.

**Manifest Hash:** Allows unambiguously identifying a bundle with all its content (for verity bundles that contain the verity hash in the manifest).

**Comment by ejoerns** [comment]
FTR: After some pre-work, partly referenced here, the base implementation for the event logging noted in this concept was submitted as PR #1230.

What is still missing in this is the actual logging of messages with the information made available in the preliminary work done in a couple of PRs (e.g. #1075, #1150, #1151, #1153)

**Comment by ejoerns** [comment]
FTR: In the meantime, with #1292 the actual logging of messages was implemented, too.

Thus, an initial evaluation and testing version of the event logging described in this concept was released with RAUC [v1.11](https://github.com/rauc/rauc/releases/tag/v1.11).
Feel free to use it and extend it. For first steps, just follow the [documentation](https://rauc.readthedocs.io/en/latest/advanced.html#installation-history-and-event-logging).

A first round of fixes was submitted with #1340 and #1343.

As a little add-on, the systemd catalog support that was shortly noted in the concept was submitted as PR #1345 and is close to getting merged.

From the concept's point-of-view, all required initial puzzle pieces were brought together and we can close this.
